src/mpid/ch4/shm/ipc/src/ipc_send.h

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#ifndef IPC_SEND_H_INCLUDED
+#define IPC_SEND_H_INCLUDED
+
+#include "ch4_impl.h"
+#include "mpidimpl.h"
+#include "shm_control.h"
+#include "ipc_pre.h"
+#include "ipc_p2p.h"
+#ifdef MPIDI_CH4_SHM_ENABLE_XPMEM
+#include "../xpmem/xpmem_pre.h"
+#endif
+
+/*
+=== BEGIN_MPI_T_CVAR_INFO_BLOCK ===
+
+cvars:
+    - name        : MPIR_CVAR_CH4_XPMEM_LMT_MSG_SIZE
+      category    : CH4
+      type        : int
+      default     : 4096
+      class       : none
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_ALL_EQ
+      description : >-
+        If a send message size is greater than or equal to MPIR_CVAR_CH4_XPMEM_LMT_MSG_SIZE (in
+        bytes), then enable XPMEM-based single copy protocol for intranode communication. The
+        environment variable is valid only when then XPMEM shmmod is enabled.
+
+=== END_MPI_T_CVAR_INFO_BLOCK ===
+*/
+
+MPL_STATIC_INLINE_PREFIX int MPIDI_IPC_mpi_isend(const void *buf, MPI_Aint count,
+                                                 MPI_Datatype datatype, int rank, int tag,
+                                                 MPIR_Comm * comm, int context_offset,
+                                                 MPIDI_av_entry_t * addr, MPIR_Request ** request)
+{
+    int mpi_errno = MPI_SUCCESS;
+    bool dt_contig;
+    size_t data_sz ATTRIBUTE((unused));
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_IPC_MPI_ISEND);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_IPC_MPI_ISEND);
+
+    MPIDI_Datatype_check_contig_size(datatype, count, dt_contig, data_sz);
+
+    if (dt_contig) {
+#ifdef MPIDI_CH4_SHM_ENABLE_XPMEM
+        if (MPIR_CVAR_CH4_XPMEM_LMT_MSG_SIZE > -1 &&
+            dt_contig && data_sz >= MPIR_CVAR_CH4_XPMEM_LMT_MSG_SIZE && rank != comm->rank) {
+            mpi_errno = MPIDI_IPC_send_contig_lmt(buf, count, datatype, rank, tag, comm,
+                                                  context_offset, addr, MPIDI_SHM_IPC_TYPE__XPMEM,
+                                                  request);
+            MPIR_ERR_CHECK(mpi_errno);
+
```

Comment:
@gcongiu and I are discussing whether we need call `MPIDI_IPC_XPMEM|GPU_mpi_isend` and let the submodule decide which generic protocol is used. I am thinking the case that GPU and XPMEM may have different ways to transfer message. E.g., GPU may want to avoid DMA read thus it is not always `sender-open-handle-receiver-copy`.

I think it is better to define a set of generic IPC protocols (e.g., contig_send_lmt, contig_recv_lmt, contig_send_lmt_coop, ...), and let the submodule choose the appropriate protocol.

@pavanbalaji can you please provide your thought on this?

Reply:
First of all, receiver-driven protocol does not mean DMA read.  So please don't confuse those two.  I don't see any major reason why DMA read would need to be used, except in minor corner cases like noncontiguous unpack.  So I don't really understand if your question is about DMA read protocols or receiver-driven protocols.  I'm going to assume it's the latter because we almost never use DMA read for GPUs.

Second, sender driven vs. receiver driven are still IPC-level protocols, not XPMEM or GPU-level protocols.  Just because there is a sender-driven protocol available at the IPC layer, the GPU submodule does not have to use it.  The IPC layer would obviously need to make a decision as to when to use each given protocol for GPUs (e.g., if the user wants to prioritize asynchronous progress, even if there's a slight loss in performance).

Reply:
In the current receiver-driven (named sender-initialized in current code) protocol, the receiver will open the remote memory handle and trigger DMA read (from remote sbuf to local rbuf), right?

Reply:
No, the receiver will open the GPU context for the source GPU and issue a DMA write from there.  Like I said, we rarely ever do DMA read.

Reply:
I see. I may misunderstand it. I think we can go with the current code format in this PR and review/revise it when necessary (e.g., if it is too messy to make all submodule-specific protocol decisions in the IPC layer).